odl/space/npy_tensors.py

```
@@ -0,0 +1,2385 @@
+
+
+def _blas_is_applicable(*args):
+    """Whether BLAS routines can be applied or not.
+
+    BLAS routines are available for single and double precision
+    float or complex data only. If the arrays are non-contiguous,
+    BLAS methods are usually slower, and array-writing routines do
+    not work at all. Hence, only contiguous arrays are allowed.
+
+    Parameters
+    ----------
+    x1,...,xN : `NumpyGeneralizedTensor`
+        The tensors to be tested for BLAS conformity.
+    """
+    if any(x.dtype != args[0].dtype for x in args[1:]):
+        return False
+    if any(x.dtype not in _BLAS_DTYPES for x in args):
+        return False
+    if not (all(x.flags.f_contiguous for x in args) or
+            all(x.flags.c_contiguous for x in args)):
+        return False
+    return True
+
+
+def _lincomb(a, x1, b, x2, out, dtype):
+    """Raw linear combination depending on data type."""
+    # Shortcut for small problems
+    # if x1.size < 100:  # small array optimization
+    #     out.data[:] = a * x1.data + b * x2.data
+    #     return
+
+    # Use blas for larger problems
+    def fallback_axpy(x1, x2, n, a):
+        """Fallback axpy implementation avoiding copy."""
+        if a != 0:
+            x2 /= a
+            x2 += x1
+            x2 *= a
+        return x2
+
+    def fallback_scal(a, x, n):
+        """Fallback scal implementation."""
+        x *= a
+        return x
+
+    def fallback_copy(x1, x2, n):
+        """Fallback copy implementation."""
+        x2[...] = x1[...]
+        return x2
+
+    # Need flat data for BLAS, otherwise in-place does not work
+    x1_arr = x1.data.ravel(order=x1.space.view_order)
+    x2_arr = x2.data.ravel(order=x1.space.view_order)
+    out_arr = out.data.ravel(order=x1.space.view_order)
+
+    if _blas_is_applicable(x1.data, x2.data, out.data):
+        axpy, scal, copy = linalg.blas.get_blas_funcs(
+            ['axpy', 'scal', 'copy'], arrays=(x1.data, x2.data, out.data))
```

More a point of consistency. You shouldn't assume what `get_blas_funcs` does. Perhaps they change it so it acts differently depending on dimension, then we dont want a silent error popping up.